// satisfactory/index.js
'use strict'

const data = require('./data.json')

function computeMachines(args) {
    const {inputAmount, recipeInputAmount, recipeOutputAmount, recipeOutputFluidAmount, perFloor, powerShards} = args,
        floors = Math.ceil(inputAmount / (recipeInputAmount * perFloor * (1 + powerShards * 0.5))),
        machines = floors * perFloor,
        clock = inputAmount / (machines * recipeInputAmount),
        usedPowerShards = clock > 1 ? Math.ceil((clock - 1) * 2) : 0,
        outputAmount = recipeOutputAmount * machines * clock,
        outputFluidAmount = recipeOutputFluidAmount * machines * clock
    return {floors, machines, clock, outputAmount, outputFluidAmount, perFloor, powerShards: usedPowerShards * machines}
}

// create comment string to display next to machine nodes
function renderMachines(machines, names) {
    const clock = Math.round(machines.clock * 10000) / 100
    let output = ''
    if (machines.perFloor > 1) {
        output = `${machines.floors} floors of ${machines.perFloor} ${names.machinesName} @ ${clock}%`
    } else {
        output = `${machines.machines} ${names.machinesName} @ ${clock}%`
    }
    if (machines.powerShards > 0) output += ` (using ${machines.powerShards} power shards)`
    if (machines.outputAmount) output += `\nProducing ${machines.outputAmount} ${names.productName}/min`
    if (machines.outputFluidAmount) output += `\nProducing ${machines.outputFluidAmount} m³/min of ${names.fluidName}`
    return output
}

// create string to display inputs
function renderInputs(inputs) {
    let result = []
    for (let slug in inputs) result.push(`${inputs[slug]} ${data.items[slug]}/min`)
    return result.join(' + ')
}
// create string to display recipe inputs
function renderRecipe(recipe) {
    let result = []
    for (let slug in recipe.ingredients) {
        let ingredient = recipe.ingredients[slug]
        result.push(`${ingredient.amount} ${ingredient.name}/min`)
    }
    return `Recipe: ${result.join(' + ')}`
}
// create string to display recipe inputs for refineries etc.
function renderFluidRecipe(recipe) {
    let result = []
    if (recipe.ingredients.items) result.push(`${recipe.ingredients.items.amount} ${recipe.ingredients.items.name}/min`)
    if (recipe.ingredients.fluid) result.push(`${recipe.ingredients.fluid.amount} m³/min of ${recipe.ingredients.fluid.name}`)
    return `Recipe: ${result.join(' + ')}`
}

// check that inputs are valid (Assemblers, Manufacturers, etc.)
function checkInputs(inputNodes, machines, recipe) {
    let inputs = {}, expectedInputs = {}
    for (let slug in recipe.ingredients) {
        expectedInputs[slug] = recipe.ingredients[slug].amount * machines.machines * machines.clock
    }
    for (let input of inputNodes) {
        if (input.value.slug) inputs[input.value.slug] = input.value.amount
    }
    const inputsString = renderInputs(inputs)
    for (let slug in expectedInputs) {
        if (inputs[slug] == expectedInputs[slug]) delete inputs[slug]
        else throw {
            error: "Invalid inputs or amount mismatch!",
            inputs: inputsString,
            expectedInputs: renderInputs(expectedInputs),
            recipe: renderRecipe(recipe)
        }
    }
}

module.exports = {data, computeMachines, renderMachines, checkInputs, renderRecipe, renderFluidRecipe}
